### PR TITLE
Implement cloud function to post Sentry events to Salesforce

### PR DESCRIPTION
**Pull request by alalamav**
* The cloud function and enclosing package follow the example of `metrics_server`.
* It receives Sentry events via webhook integration, extracts the data and posts it as form data to Salesforce.
* Infers whether the event is from development or production and posts to the corresponding Salesforce endpoint.

### CODE REVIEW

File: src/sentry_webhook/index.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  }
+  // Use the request message if SentryEvent.message is unpopulated.
+  sentryEvent.message = sentryEvent.message || req.body.message;
+  postSentryEventToSalesforce(sentryEvent, req.body.project)
+      .then(() => {
+        res.status(200).send();
+      })
+      .catch((e) => {
+        console.error(e);
+        // Send an OK response to Sentry - they don't need to know about errors with posting to
```
Comment (trevj) [inline src/sentry_webhook/index.ts]: You'll hate me for asking but did you test if Sentry will retry a webhook if it receives, e.g, a 500?

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  recordType: '0120b0000006e8i',
+};
+
+// Posts a Sentry event to Salesforce using predefined form data. Only posts if the event contains
+// an email address.
+export function postSentryEventToSalesforce(
+    event: sentry.SentryEvent, project: string): Promise<void> {
+  return new Promise((resolve, reject) => {
+    if (!event.user || !event.user.email) {
+      reject(new Error('Not posting event without email'));
```
Comment (trevj) [inline src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Seems like this way we'll get a lot of errors in our logs. How about extracting a function called something like `isSalesforceReport` and calling that first?

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  if (!tags) {
+    return null;
+  }
+  const tagsMap = new Map<string, string>();
+  for (const i of Object.keys(tags)) {
+    try {
+      const tagEntry = tags[i];
+      tagsMap.set(tagEntry[0], tagEntry[1]);
+    } catch (e) {
+      console.error('Failed to process tag entry');
```
Comment (trevj) [inline src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Don't swallow the exception.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    formFields: SalesforceFormFields, formValues: SalesforceFormValues, event: sentry.SentryEvent,
+    email: string, isClient: boolean): string {
+  const form = [];
+  form.push(encodeFormData(formFields.orgId, formValues.orgId));
+  form.push(encodeFormData(formFields.recordType, formValues.recordType));
+  form.push(encodeFormData(formFields.email, email));
+  form.push(encodeFormData(formFields.sentryEventId, event.event_id));
+  form.push(encodeFormData(formFields.description, event.message));
+  form.push(encodeFormData(formFields.type, isClient ? 'Outline client' : 'Outline manager'));
+  const tags = getTagsMap(event.tags);
```
Comment (trevj) [inline src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Strongly suggest we:
- make the argument to `getTagsMap` required
- change this to `if (event.tags) { // call getTagsMap }`

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    form.push(encodeFormData(formFields.version, tags.get('sentry:release')));
+    if (!isClient) {
+      form.push(encodeFormData(formFields.cloudProvider, tags.get('cloudProvider')));
+    }
+  }
+  return form.join('&');
+}
+
+function encodeFormData(field: string, value?: string) {
+  return `${encodeURIComponent(field)}=${encodeURIComponent(value || '')}`;
```
Comment (trevj) [inline src/sentry_webhook/post_sentry_event_to_salesforce.ts]: FYI `encodeURIComponent` doesn't throw when passed the empty string

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        (res) => {
+          if (res.statusCode === 200) {
+            resolve();
+          } else {
+            reject(new Error(`Failed to post form data, response status: ${res.statusCode}`));
+          }
+        });
+    req.on('error', (err) => {
+      reject(new Error(
+          `Failed to submit form: ${err}`,
```
Comment (trevj) [inline src/sentry_webhook/post_sentry_event_to_salesforce.ts]: the trailing comma is messing up the formatting here

File: src/sentry_webhook/index.ts (hunk 1 above)
Comment (alalamav) [reply src/sentry_webhook/index.ts]: I did, it does not retry. Yet, I believe we shouldn't be sending them errors unless there is a problem with the request itself.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 2 above)
Comment (alalamav) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Done in 4d60e67.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 3 above)
Comment (alalamav) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: I don't want to fail the whole operation for one malformed tag. I added this as a precaution, though I have not seen malformed tags in any reports.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 4 above)
Comment (alalamav) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Good suggestion. Done in 4d60e67.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 5 above)
Comment (alalamav) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: It encodes `undefined` to `"undefined"` though - rather have an empty string.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 6 above)
Comment (alalamav) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Done in 4d60e67.

File: src/sentry_webhook/index.ts (hunk 1 above)
Comment (trevj) [reply src/sentry_webhook/index.ts]: If they don't retry, I agree.

File: src/sentry_webhook/post_sentry_event_to_salesforce.ts (hunk 3 above)
Comment (trevj) [reply src/sentry_webhook/post_sentry_event_to_salesforce.ts]: Mmm, okay. Let's make it a `warn` then, and include some debugging in the message, e.g.:
```ts
console.warn(`failed to process tag entry ${tagEntry[0]}`);
```
(does the environment have `console.warn`? I forget)

**Pull request merged**